src/hooks/use-dropdown-menu.tsx

```diff
@@ -0,0 +1,152 @@
+// Imports
+import React, { useState, useRef, createRef, useEffect } from 'react';
+import useUpdateOnlyEffect from './use-update-only-effect';
+
+
+// A custom Hook that abstracts away the listeners/controls for dropdown menus
+export default (itemCount: number) => {
+   // Use state
+   const [isOpen, setIsOpen] = useState<boolean>(false);
+   const [returnFocusToButton, setReturnFocusToButton] = useState<boolean>(false);
+   const [currentFocusIndex, setCurrentFocusIndex] = useState<number | null>(null);
+   const firstRun = useRef(true);
+   
+   
+   // Create refs
+   const buttonRef = useRef<HTMLButtonElement>(null);
+   const itemRefs = useRef([...Array(itemCount)].map(() => createRef<HTMLAnchorElement>()));
+   
+   
+   // Create type guard
+   const isKeyboardEvent = (e: React.KeyboardEvent | React.MouseEvent): e is React.KeyboardEvent => (e as React.KeyboardEvent).key !== undefined;
+   
+   
+   // Handle changes in the state of the menu
+   useEffect(() => {
+       // Return if the ref currently points to a null value
+       if (!itemRefs.current[0].current || !buttonRef.current) {
+           return;
+       }
+
+
+       // Return if this is the first fire of the hook, and update ref
+       if (firstRun.current) {
+           firstRun.current = false;
+           return;
+       }
+
+       
+       // If the menu is currently open focus on the first item in the menu, else return focus to the button if state indicates that is desired behavior
+       if (isOpen) {
+           itemRefs.current[0].current.focus();
+           setCurrentFocusIndex(0);
+       } else if (returnFocusToButton) {
+           buttonRef.current.focus();
+       }
+   }, [isOpen, returnFocusToButton]);
+   
+   
+   // Handle changes in the currently focused element
+   useEffect(() => {
+       if (currentFocusIndex !== null) {
+           itemRefs.current[currentFocusIndex].current!.focus(); // TODO: Optional chaining
```

Comment:
Because this logic is contained within a `useEffect` Hook, to get the focus to change (which is simply one line of code), you have to force this component to rerender.

This seems like an indirect and inefficient way to change the focus.

Instead, do you think it'd be possible to save the current focus index in a ref rather than in state, and then simply directly call a `moveFocus` function that updates the ref's value and calls the `.focus()` method on the correct item?

Reply:
Using a ref would ensure that the correct index was retained across rerenders, but updating refs don't trigger a rerender themselves. Which should be fine, since the focus is something that can change without rerendering the component.